### ISSUE TITLE
Global Gradient calculation is turned off during validation step.

### ISSUE THREAD

**Issue by Nilanshrajput** [PyTorchLightning/pytorch-lightning#2180]
If an error occurs during the validation step, the tradition calculation is turned off for the runtime, you have to either specifically enable it or restart runtime!

**Comment by Borda** [comment]
@Nilanshrajput is it a question?

**Comment by Nilanshrajput** [comment]
@Borda, no it's a bug!

**Comment by Borda** [comment]
could you please describe more what is happening or give an example?

**Comment by Nilanshrajput** [comment]
I used the  MNIST example for demonstrating the error. https://colab.research.google.com/drive/1LWlUurpIu_pD4fP_cUgwhF5f0nS6fvb4?usp=sharing 
Here in validation step I added `assert1==2` which will cause the error during execution, to remove the error you can comment that line, but this error will pop-up `element 0 of tensors does not require grad and does not have a grad_fn` . This error is generated because gradient calculation was turned off during the validation step globally, and due to error while executing and never reverted.

**Comment by Nilanshrajput** [comment]
`# disable gradients to save memory
        torch.set_grad_enabled(False)
`
This line in trainer.evaluation_loop.py sets global grad calculation False, this is reverted at the end of evaluation loop but that's is never executed if there is an error raised in between, and the model remain in eval mode with gradient calculation disabled.

**Comment by Nilanshrajput** [comment]
@Borda This could be resolved in two ways i think:
1. use local context for disabling grad ` with torch.no_grad():`. but here also if error occurs the model will remain in eval mode, also for a long piece of code it might not look good.
2. At the start of training_loop add following lines `model.train()
        torch.set_grad_enabled(True)`. This is better method I think.
I can add a PR if you agree.

**Comment by Borda** [comment]
> 2. At the start of training_loop add following lines `model.train() torch.set_grad_enabled(True)`. This is better method I think.
>    I can add a PR if you agree.

Yeah, sending a PR would be great...
cc: @awaelchli

**Comment by williamFalcon** [comment]
If an error is raised, the program should crash no?

**Comment by Nilanshrajput** [comment]
No program should crash, but when you are working in notebooks, after resolving the error, the grad calculation does not take place. as in validation step it was turned off and never turned back on.
Can you try the collab link i provided, after resolving error you have start the runtime.
